File: kirun-js/src/engine/function/system/array/DeleteFrom.ts
import { KIRuntimeException } from '../../../exception/KIRuntimeException';
import { EventResult } from '../../../model/EventResult';
import { FunctionOutput } from '../../../model/FunctionOutput';
import { FunctionExecutionParameters } from '../../../runtime/FunctionExecutionParameters';
import { AbstractArrayFunction } from './AbstractArrayFunction';

export class DeleteFrom extends AbstractArrayFunction {
    public constructor() {
        super(
            'DeleteFrom',
            [
                DeleteFrom.PARAMETER_ARRAY_SOURCE,
                DeleteFrom.PARAMETER_INT_SOURCE_FROM,
                DeleteFrom.PARAMETER_INT_LENGTH,
            ],
            DeleteFrom.EVENT_RESULT_EMPTY,
        );
    }

    protected async internalExecute(context: FunctionExecutionParameters): Promise<FunctionOutput> {
        let source: any[] = context
            ?.getArguments()
            ?.get(DeleteFrom.PARAMETER_ARRAY_SOURCE.getParameterName());

        let start: number = context
            ?.getArguments()
            ?.get(DeleteFrom.PARAMETER_INT_SOURCE_FROM.getParameterName());

        let len: number = context
            ?.getArguments()
            ?.get(DeleteFrom.PARAMETER_INT_LENGTH.getParameterName());

        if (source.length == 0) throw new KIRuntimeException('There are no elements to be deleted');

        if (start >= source.length || start < 0)
            throw new KIRuntimeException(
                'The int source for the array should be in between 0 and length of the array ',
            );

        if (len == -1) len = source.length - start;

        if (start + len > source.length)
            throw new KIRuntimeException(
                'Requested length to be deleted is more than the size of array ',
            );

        source.splice(start, len);

        return new FunctionOutput([EventResult.outputOf(new Map([]))]);
    }
}

File: kirun-py/src/kirun_py/function/system/array/delete_from.py
from __future__ import annotations

from typing import Any, List, TYPE_CHECKING

from kirun_py.exception.ki_runtime_exception import KIRuntimeException
from kirun_py.function.system.array.abstract_array_function import AbstractArrayFunction
from kirun_py.model.event_result import EventResult
from kirun_py.model.function_output import FunctionOutput

if TYPE_CHECKING:
    from kirun_py.runtime.function_execution_parameters import FunctionExecutionParameters


class DeleteFrom(AbstractArrayFunction):

    def __init__(self) -> None:
        super().__init__(
            'DeleteFrom',
            [
                AbstractArrayFunction.PARAMETER_ARRAY_SOURCE,
                AbstractArrayFunction.PARAMETER_INT_SOURCE_FROM,
                AbstractArrayFunction.PARAMETER_INT_LENGTH,
            ],
            AbstractArrayFunction.EVENT_RESULT_ARRAY,
        )

    async def internal_execute(
        self, context: FunctionExecutionParameters
    ) -> FunctionOutput:
        source: List[Any] = list(
            context.get_arguments().get(
                AbstractArrayFunction.PARAMETER_ARRAY_SOURCE.get_parameter_name()
            )
        )
        start: int = context.get_arguments().get(
            AbstractArrayFunction.PARAMETER_INT_SOURCE_FROM.get_parameter_name()
        )
        length: int = context.get_arguments().get(
            AbstractArrayFunction.PARAMETER_INT_LENGTH.get_parameter_name()
        )

        if len(source) == 0:
            raise KIRuntimeException('There are no elements to be deleted')

        if start >= len(source) or start < 0:
            raise KIRuntimeException(
                'The int source for the array should be in between 0 and length of the array '
            )

        if length == -1:
            length = len(source) - start

        if start + length > len(source):
            raise KIRuntimeException(
                'Requested length to be deleted is more than the size of array '
            )

        del source[start:start + length]

        return FunctionOutput([
            EventResult.output_of(
                {AbstractArrayFunction.EVENT_RESULT_NAME: source}
            )
        ])
